### ISSUE TITLE
Schism fails player abuse test 25 (portamento and pitch slide)

### ISSUE THREAD

**Issue by jangler** [schismtracker/schismtracker#3]
It's been failing this test for the past seven years, apparently. Not sure why no one ever bothered to fix it. Too obscure? See https://github.com/jangler/schismtracker/wiki/Player-abuse-tests#portamento-and-pitch-slide for details on what _should_ happen. What _does_ happen in Schism is that all four notes play identically.

**Comment by jangler** [comment]
Note: when fixing this, make sure that both linear and Amiga pitch slides are accounted for.

**Comment by jangler** [comment]
OK, incredibly, in Impulse Tracker when Fx(x) and Ex(x) share effect memory with Gx(x), _the shared value isn't retrieved until the first tick of the next row of an effect with a zero parameter_. That's pretty hard to parse, so I'll draw an example here:

```
000 C-5 01 -- .00
001 C-1 01 -- G01
002 ... -- F1 G00
003 ... -- F0 G00
```

On row 000, a note plays. On row 001, a slow pitch slide down starts. On row 002, that pitch slide continues at the same rate while a fast pitch slide up simultaneously begins. This is where things get interesting.

On the first tick of row 002, the G00 effect retrieves the value of 01 from the shared effect memory, then the F1 effect sets the shared effect memory value to 04, but _the G00 continues sliding at a speed of 01 until the next row_.

It's only on row 003 that the G00 and F0 both retrieve the value of 04 from shared effect memory and cancel each other out.

Luckily, I _think_ this is the only instance where this quirk of effect memory storage can cause an incompatibility. I'm not sure what sort of maniac would write different types of pitch slides in the same channel simultaneously, but I'm still going to fix this anyway.

**Comment by jangler** [comment]
OK, I'm actually not going to fix that final case. It's extremely obscure, OpenMPT doesn't handle it "correctly" either, and according to the [player abuse tests](https://github.com/jangler/schismtracker/wiki/Player-abuse-tests#portamento-and-pitch-slide) the only player that _does_ handle it "correctly" is DUMB. I don't think it's worth complicating the effect memory implementation for the sake of fixing a bug related to simultaneous pitch sliding up and down. Using effect memory. For a single row.

**Comment by iamgreaser** [comment]
Contributing some notes, although I suspect it would be better to import IT's playroutine and tailor it to suit (once I've ironed all the bugs out of the C port, that is) in order to fix this. (Honestly, I think Hx+Hxx would be a more useful case to have working properly.)

sackit (my playroutine) handles this edge case fine, and fairly elegantly. Here's the approach:
- There's a variable called `pchn->eff_slide_pitch`, which is the actual effect memory (and is shared between Ex and Exx). You probably have something equivalient.
- There's a variable called `pchn->slide_pitch`. It is reset to 0 on the first tick. Fx/Fxx adds to it, Ex/Exx subtracts from it.
- **Voleffects are processed after regular effects.** This is the critical point.

This approach is not actually how IT does it, though, but I suspect it's functionally equivalent (but yes, IT _does_ process voleffects after regular effects).

Here's how it works in the real thing: Ex/Fx//Gx actually share the same value for performing slides (`EFG`), while Exx/Fxx/Gxx all copy the slide variable to the not-strictly-assigned-to-anything-in-particular space (`_40`).

And here's the code from libit216 (incomplete but mostly working C translation of the source-code asm release)...

Gxx:

```
void CommandG(it_engine *ite, it_host *chn)
{
    if((chn->Flags & 16) == 0)
        return;

    int16_t bx = chn->_40;

    if((chn->_42 & 0xFF) != 1)
    {
        // Slide down
        PitchSlideDown(ite, chn, slave, bx);
```

Gx volcmd:

```
void VolumeCommandG(it_engine *ite, it_host *chn)
{
    if((chn->Flags & 0x16) == 0)
        return;

    uint16_t bx = chn->EFG;
    if((ite->hdr.Flags & 0x0020) != 0) // Link command G?
        bx = chn->GOE;

    if(bx == 0)
        return;

    bx <<= 2;
    it_slave *slave = &ite->slave[chn->SCOffst];

    if(chn->_42 != 1)
    {
        // Slide down
        PitchSlideDown(ite, chn, slave, bx);
```

Exx:

```
void CommandE(it_engine *ite, it_host *chn)
{
    CommandEChain(ite, chn, chn->_40);
}

void CommandEChain(it_engine *ite, it_host *chn, int16_t bx)
{
    it_slave *slave = &ite->slave[chn->SCOffst];
    PitchSlideDown(ite, chn, slave, bx);
    slave->Frequency_Set = slave->Frequency;
}
```

Ex volcmd:

```
void VolumeCommandE(it_engine *ite, it_host *chn) // Pitch slide down
{
    CommandEChain(ite, chn, ((int16_t)(uint16_t)chn->EFG)<<2);
}
```

**Comment by jangler** [comment]
Yeah, we're processing volume columns effects after effect column effects (bizarrely, this was not the case when we originally imported the hg repo, despite mention of that being the correct behavior in the [player abuse test notes](https://github.com/jangler/schismtracker/wiki/Player-abuse-tests#portamento-and-pitch-slide)).

> - There's a variable called pchn->slide_pitch. It is reset to 0 on the first tick. Fx/Fxx adds to it, Ex/Exx subtracts from it.

That's what we're missing, and why Schism plays the first note of 25.it incorrectly. I talked with sagamusix about this issue briefly, and he's aware of it as well but agrees that it's not worth fixing in OpenMPT either. If you want to make a PR that fixes the issue without complicating the current code too much, by all means go ahead! The relevant code is in [player/effects.c](https://github.com/jangler/schismtracker/blob/master/player/effects.c).

We're not ready to completely replace the playback logic at this point, although this might become a possibility in the future if we keep running into incompatibilities that conflict with the current structure of the playback code.

**Comment by sagamusix** [comment]
For the record, this was fixed in 223e327d9448561931b8cac8a55180286b17276c (and in OpenMPT of course) as I found myself accidentally abusing this quirk in several tracks (and thus they play incorrectly in IT, but even XMPlay played them the same way as OpenMPT / Schism Tracker).

(plot twist: volume commands *do* run before regular commands in Impulse Tracker. That wasn't the problem here. The main issue is that IT evaluates command parameters at the start of the row and then never again. Details in https://wiki.openmpt.org/Development:_Test_Cases/IT#DoubleSlide.it)
